krabcake/rs_hello/src/lib.rs

```diff
@@ -79,3 +111,939 @@ fn rust_eh_personality() {
 // pnkfelix got tired of fighting with the linker
 // "if you cannot beat 'em, join 'em."
 mod libc_stuff;
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+struct Tag(u64);
+
+impl Tag {
+    fn next(self) -> Tag {
+        Tag(self.0 + 1)
+    }
+}
+
+static mut COUNTER: Tag = Tag(0);
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+enum Item {
+    Unique(Tag),
+}
+
+impl Item {
+    fn num(&self) -> u64 {
+        match *self {
+            Item::Unique(Tag(val)) => val,
+        }
+    }
+}
+
+// FIXME: Need to more clearly distinguish between addresses and
+// shadowing of values. I.e. the tag here needs to be attached to the
+// location where an address itself is being stored, and needs to flow
+// along with the address as it is copied from place to place.
+static mut TRACKED_ADDRS: Vec<(vg_addr, Tag)> = Vec::new();
+static mut TRACKED_TEMPS: Vec<(vg_uint, Tag)> = Vec::new();
+// "GREG" is "Guest Register", i.e. the emulated registers.
+static mut TRACKED_GREGS: Vec<(vg_ulong, Tag)> = Vec::new();
+
+static mut STACKS: Vec<(vg_addr, Vec<Item>)> = Vec::new();
+
+// FIXME UGLY HACK
+// this is a big bad hack: I haven't managed to connect up
+// the client request return point into the rest of the
+// data-flow of the krabcake-rewritten code. So for now,
+// use a one-length event queue that each memory operation
+// that I *have* instrumented will check, and will process
+// the relevant event if it sees the corresponding memory
+// location.
+
+type SbEvent = (SbEventKind, vg_addr, Tag);
+static mut STACKED_BORROW_EVENT: Option<SbEvent> = None;
+
+macro_rules! msg {
+    ($x: literal) => { msg!($x,) };
+    ($x: literal, $($arg: expr $(,)?),*) => {{
+        let x: &[u8] = $x;
+        assert!(x.last() == Some(&b'\0'));
+        if false { vgPlain_printf(x.as_ptr() as *const c_char, $($arg),*); }
+    }};
+}
+
+// This is the same as msg for now; I just wanted an easy way
+// to disable the msg's in the future while keeping the alerts.
+macro_rules! alert {
+    ($x: literal) => { msg!($x,) };
+    ($x: literal, $($arg: expr $(,)?),*) => {{
+        let x: &[u8] = $x;
+        assert!(x.last() == Some(&b'\0'));
+        vgPlain_umsg(x.as_ptr() as *const c_char, $($arg),*);
+    }};
+}
+
+unsafe fn print_stacked_borrow_event(event: SbEvent) {
+    msg!(
+        b"SbEvent(kind: %s, addr: 0x%08llx, tag: %d)\0",
+        (event.0).c_str(),
+        event.1,
+        (event.2).0,
+    );
+}
+
+unsafe fn if_sb_event_queued_print(before: impl FnOnce(), after: impl FnOnce()) {
+    if let Some(event) = STACKED_BORROW_EVENT {
+        before();
+        print_stacked_borrow_event(event);
+        after();
+    }
+}
+
+#[derive(Copy, Clone)]
+enum SbEventKind {
+    BorrowMut = 0x2000,
```

Comment:
same comment as above; where does this number come from?

Reply:
You are right, **this** number definitely is shared among multiple files and needs proper referencing.

(or I need to figure out how to pull in the other crate that you made for this here.)